Interruptible APIs for key agreement use cases

athoelke opened the pull request:
Draft PR including the interruptible operations for `psa_generate_key(), `psa_export_public_key()` and `psa_key_agreement()`. Currently this is just the additional API elements, update to the *Functionality* chapter to cover this simpler interruptible pattern will follow in additional commits.

This is based on top of #107. The first unique commit is https://github.com/ARM-software/psa-api/commit/57ae95c9dd6cd5aed55806d1aca3c021fc5dd8e7.

Fixes #198

MarcusJGStreets reviewed (approved):
Look good to me

gilles-peskine-arm commented on doc/crypto/api/ops/key-agreement.rst:
```diff
@@ -332,6 +338,264 @@ Combining key agreement and key derivation
+        The standalone key agreement algorithm to compute: a value of type `psa_algorithm_t` such that :code:`PSA_ALG_IS_STANDALONE_KEY_AGREEMENT(alg)` is true.
+    .. param:: const psa_key_attributes_t * attributes
+        The attributes for the new key.
+        This function uses the attributes as follows:
```
The bullet points below actually mostly explain how the resulting key is constructed rather than how the `attributes` parameter is used. I find this presentation confusing. In particular, the key size is always the same regardless of the attributes. Separately, the key size in the attributes must be either correct or 0.

athoelke replied:
This text is identical to that for `psa_key_agreement()`, which we discussed at some length as part of https://github.com/ARM-software/psa-api/pull/101 (specifically commit https://github.com/ARM-software/psa-api/pull/101/commits/5c48afe702061b077754cf60156b4b1972713742).

Is there a good reason to do something different here, or to change what we have done in the existing APIs?

gilles-peskine-arm replied:
Sorry, I missed that earlier. My objection isn't specific to this function. The explanations of how the attributes are used and how the new key is constructed are confusing.

athoelke replied:
Proposal for improvement raised as #205. Will feed results of that PR into this one.

athoelke replied:
Clarifications in the updated PR #205 have been copied into this PR

gilles-peskine-arm commented on doc/crypto/api/ops/key-agreement.rst:
```diff
@@ -332,6 +338,264 @@ Combining key agreement and key derivation
+
+        *   The key permitted-algorithm policy is required for keys that will be used for a cryptographic operation, see :secref:`permitted-algorithms`.
+        *   The key usage flags define what operations are permitted with the key, see :secref:`key-usage-flags`.
+        *   The key lifetime and identifier are required for a persistent key.
```
The lifetime and identifier are also required for a volatile key, they just use different sets of values.

athoelke replied:
That text is taken from _all_ the other key creation functions, and was first used when we clarified the use of attributes 4 years ago. (Private link to original source commit: https://github.com/ARM-software/psa-crypto-api/commit/5d12b9abdcb1cc183a0ba8113771a034ecc6fd64).

Perhaps it would be clearer to state that the lifetime and identifier have to be set explicitly for a persistent key, as the default values will create a volatile key?

athoelke replied:
Proposal for improvement raised as #205

gilles-peskine-arm commented on doc/crypto/api/keys/management.rst:
```diff
@@ -529,3 +536,456 @@ Key export
+
+        The following attributes are required for all keys:
+
+        *   The key type. It must not be an asymmetric public key.
```
This suggests that implementations that support interruptible operations should support interruptible key generation for symmetric keys, even though operations _on_ symmetric keys don't have interruptible APIs. I think we should say that only key-pair types are expected to be supported, and implementations are expected to reject symmetric keys but they may accept them.

gilles-peskine-arm replied:
Although on second thoughts, it's probably one less line of code to just systematically fall back to `psa_generate_key` for any key type that doesn't have interruptible support, and let it decide between public keys, key pairs with single-iop generation, and symmetric keys.

athoelke replied:
It might be useful to state that an implementation is permitted to only support the use of this API for a subset of the key types that can be generated using `psa_generate_key()`?

athoelke replied:
I've added a commit to note that implementations might only provide incremental generation for some key types.

athoelke commented:
Rebased on main

gilles-peskine-arm commented on doc/crypto/api/ops/key-agreement.rst:
```diff
@@ -350,6 +356,276 @@ Combining key agreement and key derivation
+        Number of *ops* that the operation has taken so far.
+
+    After the interruptible operation has completed, the returned value is the number of *ops* required for the entire operation.
+    The value is reset to zero by a call to either `psa_key_agreement_iop_setup()` or `psa_key_agreement_iop_abort()`.
```
When we were discussing the specification, we implicitly assumed that setup takes negligible time. But this is not a safe assumption. For example, for key agreement, the public key needs to be validated. In Mbed TLS, we do this during setup (and it's the natural way to do it). Mbed TLS estimates the public key validation to take 3 ops. (The key agreement as a whole takes thousands of ops.)

So I think we should specify that setup resets the op count to 0 at the beginning of its operation, but then adds the number of ops that it takes. And recommend to implementations that this number should be lower than the value set with `psa_iop_set_max_ops`.

gilles-peskine-arm replied:
This applies to all the functions, although key agreement might be the only one for which it's really natural to have setup take a few ops.

athoelke replied:
That does complicate the way we can describe the reset of the _ops_ value, but is a valid point. Is it sufficient to expect that validation will occur during setup, or should the recommendation be that if validation can be very complex (substantially more _ops_ than a reasonable application-configured maximum) it should be deferred to the completion function?

Similar to the case for deferring key id allocation (`PSA_ERROR_ALREADY_EXISTS` returned by `psa_key_agreement_iop_complete()`), should we also permit `psa_key_agreement_iop_complete()` to return status codes related to the validation of the peer key? - which enables deferred validation.

gilles-peskine-arm replied:
I think we should recommend that setup should only perform a small number of ops, preferably less than whatever was set by `psa_iop_set_max_ops()`.

And I hadn't thought of that, but indeed we should allow complete() to return status codes related to validations that may be considered slow.

athoelke commented:
Rebased, and applied documentation improvements that have been merged to `main`.

athoelke commented:
Addressed the comment (https://github.com/ARM-software/psa-api/pull/107#discussion_r1882310284) in #107, as this PR has an updated version of that chapter, and the comment applies to additional parts of the text.
